source/opt/fold_spec_constant_op_and_composite_pass.cpp

```diff
@@ -316,6 +316,66 @@ bool IsValidTypeForComponentWiseOperation(const analysis::Type* type) {
   }
   return false;
 }
+
+// Encodes the integer |value| of in a word vector format appropriate for
+// representing this value as a operands for a constant definition. Performs
+// zero-extension/sign-extension/truncation when needed, based on the signess of
+// the given target type.
+//
+// Note: type |type| argument must be either Integer or Bool.
+utils::SmallVector<uint32_t, 2> EncodeIntegerAsWords(const analysis::Type& type,
+                                                     uint32_t value) {
+  const uint32_t all_ones = ~0;
+  uint32_t bit_width = 0;
+  uint32_t pad_value = 0;
+  bool result_type_signed = false;
+  if (auto* int_ty = type.AsInteger()) {
+    bit_width = int_ty->width();
+    result_type_signed = int_ty->IsSigned();
+    if (result_type_signed && static_cast<int32_t>(value) < 0) {
+      pad_value = all_ones;
+    }
+  } else if (type.AsBool()) {
+    bit_width = 1;
+  } else {
+    assert(false && "type must be Integer or Bool");
+  }
+
+  assert(bit_width > 0);
+  uint32_t first_word = value;
+  const uint32_t bits_per_word = 32;
+
+  // Truncate first_word if the |type| has width less than uint32.
+  if (bit_width < bits_per_word) {
+    const uint32_t num_high_bits_to_mask = bits_per_word - bit_width;
```

Comment:
This section is a little confusing.  It might be clearer if you just has some utility functions.  You also so extra work.  For example, you clear the bits even if you will later set all of them.  Another is fall through and doing the loop at the end.  I would try to get this down to something like:

```
if (result_type_signed && IsNegativeValue(value, bit_width)) { return { SetBits(value, 32 - bit_width); } }
else { return { ClearBits(value, 32 - bit_width); }}
```

I removes the need for many of the comments because the function names tell us what is happening.

Reply:
Yeah, I don't like bit manipulations either. Does it make more sense add these helper functions to this file or to `source/util/bitutils.h`? Not sure whether other (/future) code will need them.

Reply:
Yes, bitutil is a good place for them.